src/Media/Ingester/SideloadDir.php

```diff
@@ -0,0 +1,409 @@
+<?php declare(strict_types=1);
+
+namespace FileSideload\Media\Ingester;
+
+use Omeka\Api\Request;
+use Omeka\Entity\Media;
+use Omeka\File\TempFileFactory;
+use Omeka\File\Validator;
+use Omeka\Media\Ingester\IngesterInterface;
+use Omeka\Stdlib\ErrorStore;
+use Omeka\Stdlib\Message;
+use Laminas\Form\Element;
+use Laminas\View\Renderer\PhpRenderer;
+
+/**
+ * @todo Factorize with Sideload and avoid to list files and directory in main directory twice.
+ */
+class SideloadDir implements IngesterInterface
+{
+    /**
+     * @var string
+     */
+    protected $directory;
+
+    /**
+     * @var bool
+     */
+    protected $deleteFile;
+
+    /**
+     * @var int
+     */
+    protected $maxFiles;
+
+    /**
+     * @var TempFileFactory
+     */
+    protected $tempFileFactory;
+
+    /**
+     * @var Validator
+     */
+    protected $validator;
+
+    /**
+     * @var array
+     */
+    protected $listDirs = [];
+
+    /**
+     * @var bool
+     */
+    protected $hasMoreDirs = false;
+
+    /**
+     * @param string $directory
+     * @param bool $deleteFile
+     * @param int $maxFiles
+     * @param TempFileFactory $tempFileFactory
+     * @param Validator $validator
+     */
+    public function __construct(
+        $directory,
+        $deleteFile,
+        $maxFiles,
+        TempFileFactory $tempFileFactory,
+        Validator $validator
+    ) {
+        // Only work on the resolved real directory path.
+        $this->directory = $directory ? realpath($directory) : '';
+        $this->deleteFile = $deleteFile;
+        $this->maxFiles = $maxFiles;
+        $this->tempFileFactory = $tempFileFactory;
+        $this->validator = $validator;
+    }
+
+    public function getLabel()
+    {
+        return 'Sideload folder'; // @translate
+    }
+
+    public function getRenderer()
+    {
+        return 'file';
+    }
+
+    /**
+     * Ingest from a folder on the server.
+     *
+     * Accepts the following non-prefixed keys:
+     * - ingest_folder: (required) The source folder where the file to ingest is.
+     * - ingest_filename: (required) The filename to ingest.
+     * - ingest_folder_recursively: (optional, default false) Ingest folder recursively?
+     * - store_original: (optional, default true) Store the original file?
+     *
+     * {@inheritDoc}
+     */
+    public function ingest(Media $media, Request $request, ErrorStore $errorStore)
+    {
+        $data = $request->getContent();
+
+        // Checks are already done during pre-hydration, but another check is
+        // needed when the ingester is called directly.
```

Comment:
Are these checks necessary? As you mention, they're already done during pre-hydration but when would the ingester be called directly?